src/main/java/nl/windesheim/codeparser/analyzers/strategy/StrategyAnalyzer.java

```diff
@@ -0,0 +1,186 @@
+
+        //Create visitors which will find classes with special properties
+        EligibleStrategyContextFinder strategyContextFinder = new EligibleStrategyContextFinder();
+        ImplementationFinder implementationFinder = new ImplementationFinder();
+
+        //Without a type solver the strategy analyzer can't function
+        if (typeSolver == null) {
+            return patterns;
+        }
+
+        ArrayList<Pair<VariableDeclarator, ClassOrInterfaceDeclaration>> eligibleContexts = new ArrayList<>();
+
+        //For each file
+        for (CompilationUnit compilationUnit : files) {
+            //Reset the state of the visitor
+            strategyContextFinder.reset();
+
+            //visit all nodes and save eligible contexts in buffer
+            strategyContextFinder.visit(compilationUnit, typeSolver);
+
+            //Read buffer into collection
+            eligibleContexts.addAll(strategyContextFinder.getClasses());
+        }
+
+        //foreach eligible context class
+        for (Pair<VariableDeclarator, ClassOrInterfaceDeclaration> eligibleContext : eligibleContexts) {
+            //Get the strategy interface of which the context type has a variable
+            ClassOrInterfaceDeclaration strategyInterface = eligibleContext.getValue();
+
+            //Get the variable deceleration
+            VariableDeclarator strategyVariable = eligibleContext.getKey();
+            ClassOrInterfaceType strategyInterfaceType = (ClassOrInterfaceType) strategyVariable.getType();
+
+            //Walk up the tree until we have the class containing the variable deceleration, this is the context class
+            Node currentNode = strategyVariable;
+            while (!(currentNode instanceof ClassOrInterfaceDeclaration)) {
+                if (!currentNode.getParentNode().isPresent()) {
+                    break;
+                }
+                currentNode = currentNode.getParentNode().get();
+            }
+
+            //If we stopped but the current node is not a class we have a issue, so continue loop
+            if (!(currentNode instanceof ClassOrInterfaceDeclaration)) {
+                continue;
+            }
+
+            ClassOrInterfaceDeclaration context = (ClassOrInterfaceDeclaration) currentNode;
+
+            //Find all method calls in the context class
+            List<MethodCallExpr> methodCalls = context.findAll(MethodCallExpr.class);
+
+            boolean classCallsStrategy = false;
+
+            //Loop over the method calls
+            for (MethodCallExpr methodCallExpr : methodCalls) {
+
+                //We are looking for a call in the scope of the variable deceleration
+                if (methodCallExpr.getScope().isPresent()) {
+
+                    //Resolve the type of the scope
+                    ResolvedType scopeType = javaSymbolSolver.calculateType(methodCallExpr.getScope().get());
+                    if (scopeType instanceof ReferenceTypeImpl) {
```

Comment:
Ik zie hier 3 geneste ifjes. Graag proberen te voorkomen.

Reply:
Dit kun je bijvoorbeeld doen door de if zo te schrijven dat hij eruit klapt als de situatie niet zo is als hij zou moeten zijn ipv checken of hij wel zo is.